Clear 'item' in Autocomplete widget when textbox cleared in IE

vansimke opened the pull request:
This was originally requested to be pulled under [PR#118](https://github.com/dojo/dijit/pull/118), which was inadvertently closed. In this PR, the new functionality has been added to the _onInput method of the dijit/form/_AutoCompleterMixin per @wkeese's recommendation in the original PR.

wkeese commented:
I see you added 7 commits about this to dijit already.   I guess they all cancel each other out?

This code looks better than before, but I have a few questions:

1) Why do we need special handling for when the `<input>` becomes empty?  IIRC FilteringSelect already has code to react to any change in the value of the `<input>`.

2) Is the setTimeout() really needed, and if so why?

3) In the existing code (i.e. without any of your changes), I modified _autoComplete.html so that the first option is `<option value="AL"></option>` instead of `<option value="AL">Alabama</option>`   I then opened test_FilteringSelect.html, and tried various ways of selecting that first option (by mouse, or by using the keyboard to backspace and clear the field).   It correctly set item to "AL".  Doesn't your change break that case?

vansimke commented:
Hello @wkeese,

I'm not going to have time to get an update put together, but I did want to get back to you. 

1) IE has a 'clear' button that it applies to textboxes. When pressed, it will clear the elements contents, but that isn't being picked up by dijit/form/FilteringSelect, so the 'item' property is not being updated properly. (as documented in [18862](https://bugs.dojotoolkit.org/ticket/18862))

2) No it's not. This was a holdover from when I was detecting 'mouseup' in my original implementation. It isn't required when listening to the input event. 

3) You're correct on this, and I hadn't considered this case. This is actually what is going to take a bit more time to figure out a plan. I'll keep noodling on it and

wkeese commented:
OK, but what I'm asking is *why* dijit/form/FilteringSelect isn't picking up when IE clears the field.   It doesn't sound like it needs to be a special case.   Obviously, it's not working, hence the original bug report, but I wonder why dijit/form/FilteringSelect handles any change to the `<input>` except for it being completely cleared.

vansimke commented:
Okay, I think I've found the answer to you're (really good) question. It turns out that the event IS being caught, but the charOrCode of the event object is not included in the event object. That causes the [_processInput](https://github.com/dojo/dijit/blob/master/form/_SearchMixin.js#L128) method in the _SearchMixin to bypass the search, since the code has to be a string or code '229'. 

Given this, we could just detect that condition (typeof evt.charOrCode === 'undefined'), but that had the potential of triggering extraneous searches if there are other input events that behave the same way. 

@wkeese do you have opinion on how to proceed?

wkeese commented:
@vansimke thanks for tracking that down but of course, it brings up more questions (and by the way, sorry this is so complicated).

I see the code in _SearchMixin#_processInput() to look at `evt.charOrCode`, but it seems strange, because there are multiple cases where `evt.charOrCode` isn't set.   Besides the issue where the user clicks the [x] to clear the `<input>`, there's also cut and paste, and I'm sure we tested that before.   Interestingly, the description of _SearchMixin#_processInput() specifically says that it handles cut-and-paste.

So, I think we should figure out how cut-and-paste is working (or if it is also broken).   It's true that we could check if   `typeof evt.charOrCode === 'undefined'`, or alternately to avoid extraneous searches we could just check if the `<input>`'s value has changed since last time we searched.

vansimke commented:
I experimented with cut and paste operations. It turns out that they do populate the 'value' of the evt.target with the pasted value, so no special handling appeared to be required. I agree with the idea of comparing the received value in the event with the _lastInput property. I've added that to the PR.

wkeese commented:
> I experimented with cut and paste operations. It turns out that they do populate the 'value' of the evt.target with the pasted value, so no special handling appeared to be required. 

OK... and similarly doesn't the [x] clear button set evt.target.value too?   (Specifically setting it to "".)  I don't see why [x] needs to be handled differently than a cut of all the text.

vansimke commented:
Sorry, I got my wires crossed. You were asking how cut and paste worked with an empty charOrCode field (not evt.target.value). I didn't reread that right when I came back to this issue. The answer to that is that the charOrCode for a paste event is 229 which is called out as a specific case in the relevant logical test (https://github.com/dojo/dijit/blob/master/form/_SearchMixin.js#L149). For the [x] clear button, no charOrCode is passed, which is causing that same test to set the doSearch variable to false hence, preventing the searching from being performed.

wkeese commented:
Oh I was thinking 229 represented IME input.  So how about just modifying _SearchMixin#_processInput() to not depend on `evt.charOrCode` so much, and instead to look at the `<input>`'s value?   Sounds like that would avoid the need for special case code.

vansimke commented:
That's a great idea and I think it will work. As I've poked around the _SearchMixin a bit more, I'm starting to wonder if the _processInput method needs to be as sophisticated as it is. It seems to be serving as a guard to protect against multiple queries, but the _startSearch method seems to have its own guard to ensure that the previous and current queries are not the same. Doesn't that mean that we can just turn the _processInput method into a pass-through to _startSearchFromInput and let the _startSearch method handle things?

wkeese commented:
I agree, _processInput() method can essentially be a pass-through to _startSearchFromInput(), since _startSearch() guards against redundant searches... the only tricky part is that _prev_key_backspace property, which controls whether or not auto-complete occurs on an empty input.

vansimke commented:
I've cleaned up the commit to remove the _onInput method that is no longer needed in dijit/form/_AutoCompleterMixin and cleaned out the redundant guard code from dijit/form/_SearchMixin#_processInput. @wkeese what do you think?

wkeese commented:
Thanks, at a glance it looks good.  I don't know where `_maskValidSubsetError` ever gets set to false though.

vansimke commented:
I searched all of dijit/form/** and found something here in dijit/form/ValidationTextBox: https://github.com/dojo/dijit/blob/master/form/ValidationTextBox.js#L170-L178

wkeese commented:
OK sure.   I guess that part of the code is the same before and after your change, so maybe it wasn't a fair question anyway.

wkeese commented:
I ran the regression test on IE10, and it passed, so this looks good to me to check in.   Don't forget to squash the three commits into one before pushing to this repository, with the commit comment including "refs #????" (whatever the bug number is).  Thanks!

vansimke commented:
Closed via https://github.com/dojo/dijit/commit/bf69956bfca11183aaf6e5d131985cc8e2a2a526
